lib/AST/Type.cpp

```diff
@@ -792,6 +792,20 @@ Type TypeBase::replaceCovariantResultType(Type newResultType,
       return OptionalType::get(
           objectType->replaceCovariantResultType(newResultType, uncurryLevel));
     }
+    // SWIFT_ENABLE_TENSORFLOW
```

Comment:
Note: this added logic in `TypeBase::replaceCovariantResultType` is hacky. It's necessary for `TypeConverter::getConstantOverrideInfo` (derivative vtable entry SILGen) to work properly.

I can't think of a more robust alternative that doesn't involve ~60 lines of code dupe.

---

One alternative is to drop support for `@differentiable` initializers in non-final classes.

Initializers return a dynamic `Self` type, so perhaps non-final initializer derivatives shouldn't be supported (since derivatives would return a tuple `(Self, ...)` and type-checking accepts covariant `Self` only as a top-level method result type).

If there are no use cases for `@differentiable` initializers in non-final classes, I think dropping support is fine.

Reply:
I believe we should ban `@differentiable` on everything that produces a covariant result.